File: data/src/main/java/com/ykb/app/cryptotrader/data/model/RequestUri.java
package com.ykb.app.cryptotrader.data.model;

import jakarta.persistence.*;
import lombok.*;
import org.springframework.http.HttpMethod;

import java.io.Serializable;

@Entity
@Table(name = "REQUEST_URI")
@EqualsAndHashCode(callSuper = true)
public class RequestUri extends BaseEntity {

    @EmbeddedId
    private final Key id;

    @Column(name = "IS_SECURE", nullable = false)
    @Getter
    @Setter
    private boolean isSecure;

    @Column(name = "IS_VIEW", nullable = false)
    @Getter
    @Setter
    private boolean isView;

    public RequestUri(String uri, HttpMethod httpMethod, boolean isSecure, boolean isView) {
        id = new Key(uri, httpMethod.name());
        this.isSecure = isSecure;
        this.isView = isView;
    }

    @Embeddable
    @AllArgsConstructor
    @Data
    public static class Key implements Serializable {
        @Column(name = "URI", nullable = false)
        private String uri;

        @Column(name = "HTTP_METHOD", nullable = false)
        private String httpMethod;
    }

    public void setUri(String uri) {
        id.uri = uri;
    }

    public String getUri() {
        return id.uri;
    }

    public void setServiceName(HttpMethod httpMethod) {
        id.httpMethod = httpMethod.name();
    }

    public HttpMethod getHttpMethod() {
        return HttpMethod.valueOf(id.httpMethod);
    }

}

File: library/data/data-auth/src/main/kotlin/com/ykb/app/cryptotrader/data/auth/model/RequestUri.kt
package com.ykb.app.cryptotrader.data.auth.model

import jakarta.persistence.*
import org.springframework.http.HttpMethod
import java.io.Serializable
import java.util.*

@Entity
@Table(name = "REQUEST_URI")
class RequestUri(
    serviceName: String,
    uri: String,
    httpMethod: HttpMethod,

    @Column(name = "IS_SECURE", nullable = false)
    private val isSecure: Boolean,

    @Column(name = "IS_VIEW", nullable = false)
    private val isView: Boolean
) {

    @EmbeddedId
    private val id: Key = Key(serviceName, uri, httpMethod.name())

    @Embeddable
    data class Key(
        @Column(name = "SERVICE_NAME", nullable = false)
        val serviceName: String,
        @Column(name = "URI", nullable = false)
        val uri: String,
        @Column(name = "HTTP_METHOD", nullable = false)
        val httpMethod: String
    ): Serializable

    fun getServiceName(): String {
        return id.serviceName
    }

    fun getUri(): String {
        return id.uri
    }

    fun getHttpMethod(): HttpMethod {
        return HttpMethod.valueOf(id.httpMethod)
    }

    fun isSecure(): Boolean {
        return isSecure
    }

    fun isView(): Boolean {
        return isView
    }

    override fun equals(other: Any?): Boolean {
        return other is RequestUri &&
                other.getUri() == getUri() &&
                other.getHttpMethod() == getHttpMethod() &&
                other.isSecure() == isSecure() &&
                other.isView() == isView()
    }

    override fun hashCode(): Int {
        return Objects.hash(getUri(), getHttpMethod(), isSecure(), isView())
    }

}